dom.bs

```diff
@@ -2652,6 +2652,63 @@ of a <var>node</var> into a <var>parent</var> before a <var>child</var>, run the
  <!-- Technically this is post-insert. -->
 </ol>
 
+<p>To <dfn export for=Node id=concept-node-ensure-pre-move-validity>ensure pre-move validity</dfn>
+of a <var>node</var> into a <var>parent</var> before a <var>child</var>, run these steps:
+
+<ol>
+ <li>
+  <p>If any of the following conditions are true</p>
+
+  <ul>
+   <li><p><var>parent</var> is <a>connected</a> and <var>node</var> is not <a>connected</a>; or</p></li>
+
+   <li><p><var>parent</var> is not <a>connected</a> and <var>node</var> is <a>connected</a>,</p></li>
+  </ul>
```

Comment:
Why do we need this when we have " If parent ’s shadow-including root is not the same as node ’s shadow-including root " check below?

Reply:
So, I think back when we only supported the connected->connected case, the "shadow-including root matches" check was right, because "shadow-including root" was essentially the same as "node document", in the connected case.

But now that we support the disconnected->disconnected case, I actually think the "shadow-including root" check is too strict, right? For example, imagine:

```js
const divA = document.createElement('div');
const p = divA.appendChild(document.createElement('p'));
const divB = document.createElement('div');

// Throws because divB's shadow-including root does not match p's, since they're both disconnected.
divB.moveBefore(p, null);
```

That is, in the disconnected->disconnected case, it is way more likely that the target node and destination container's shadow-including roots will not match, since in the disconnected case, a shadow-including root just tops-out at the nearest disconnected parent. So my proposal is to simply change the shadow-including root comparison to instead compare node documents. I'll do that now, and please let me know if you think that sounds good.

Reply:
Well, I don't want to support move between otherwise unrelated subtrees. This all becomes simpler if we don't have that. No need to check same document or anything like that.

Reply:
> Well, I don't want to support move between otherwise unrelated subtrees.

What are "otherwise unrelated subtrees"? We certainly want to support the code example I wrote above, no?

Reply:
I think Olli always wants there to be a shared ancestor.

Reply:
Yeah, I personally don’t think I see the value in enforcing such a requirement. Do you feel strongly about this? I think folks should be able to create detached DocumentFragments and move things around from them to possibly other detached subtrees, and mostly I don’t see the value of enforcing a common ancestor between old and new parent. What do you think @annevk ?

Reply:
I'm not sure I hold a strong opinion, but I do like the symmetry Olli is suggesting.

Reply:
Hmm, I'm not sure it provides any more symmetry than the same-document check being enforced in the connected and disconnected case. But yeah I'll just replace this check with the shadow-inclusive root check to unblock this PR, and we'll update the tests.

Reply:
Done in https://github.com/whatwg/dom/pull/1307/commits/65b4e13ade1039eee8e49316b3728e9f4191c414.